### PR TITLE
Add action to export code scanning alert to a SARIF file

### PR DESCRIPTION
**Pull request by alcaeus**
The new code-scanning-export action allows exporting open and dismissed code scanning alert to a SARIF file, which we can then submit as a release artifact. The action makes a best effort to fill in tool information along with the rules, and adds `suppression` elements to all dismissed alerts.

To test this action, you can use the following two steps in a GitHub Actions job:
```yaml
- name: "Export SARIF file from code scanning alerts"
  uses: "alcaeus/drivers-github-tools/code-scanning-export@export-code-scanning-report"

- name: "Attach generated SARIF file to build artifacts"
  uses: actions/upload-artifact@v4
  with:
    name: code-scanning-alerts.json
    path: code-scanning-alerts.json
    retention-days: 3
```

In my case, I ran the steps directly after uploading the SARIF file from an external tool. Note that the tool can only produce results if analysis of the code scanning report has completed (which is the default when uploading a SARIF file to Code Scanning). In case of using CodeQL in GitHub Actions, add the two steps after running the `github/codeql-action/analyze` action.

If the artifact is successfully uploaded, you can use a validator (e.g. [Microsoft's SARIF Validator](https://sarifweb.azurewebsites.net/Validation)) to validate that the file is valid. If the action failed or if the generated SARIF report is invalid, please report back so we can add the responsible alert to our test suite.

Note that in my case the validator complained about the tool not having a `informationUri` property. This is not exposed in our alerts, but it may be in yours in which case we can add it to the report.

TODOs:
- [ ] `npm run lint` currently reports errors (see CI pipeline)
- [ ] Update the README
- [ ] Test with some other tools and add tests for their reports

### CODE REVIEW

**Comment by blink1073** [comment]
Should the `dist` folder be added to `.gitignore`?

**Comment by alcaeus** [comment]
The `dist` folder is needed to actually run the GitHub action - maybe it should just be removed from the `pre_commit` check.

**Comment by blink1073** [comment]
> maybe it should just be removed from the pre_commit check.

SGTM

**Comment by blink1073** [comment]
Can you point to a run where you're testing this?  I tried with winkerberos and it doesn't seem to be finding the ref I just uploaded.

https://github.com/blink1073/winkerberos/pull/26
https://github.com/blink1073/winkerberos/actions/runs/9270093082/job/25502292926?pr=26

**Comment by baileympearson** [comment]
I'm seeing similar behavior.  It seems like, even though the repo has dismissed codeql alerts, the alerts are not returned by the github API.

Example when I specify `ref: main`: https://github.com/mongodb/node-mongodb-native/actions/runs/9276917785/job/25525053266?pr=4124
Example when I use the current build: https://github.com/mongodb/node-mongodb-native/actions/runs/9276879879/job/25524932613?pr=4124

(I `cat` the file in the `asdf` step)

I'm going to test further tomorrow on a fork, I think it may work if I grab all codeql alerts for a given release tag instead of the current PR.

File: code-scanning-export/src/main.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    repositoryInfo.owner,
+    repositoryInfo.repo,
+    ref,
+    core.getInput('token')
+  )
+
+  core.debug(`Found ${alerts.length} alerts, processing now...`)
+
+  const sarifReport = createSarifReport(alerts)
+  const filePath = path.join(process.cwd(), core.getInput('output-file'))
```
Comment (baileympearson) [inline code-scanning-export/src/main.ts:35]:
```suggestion
  const filePath = path.isAbsolute(filePath) ? filePath : path.join(process.cwd(), core.getInput('output-file'))
```

The shared "setup" action sets up a directory in `/tmp` for release artifacts.  I'd like to be able to provide `$S3_ASSETS/sarif-report.json` (which expands to `/tmp/s3assets/sarif-report.json` to this action, and have the generated sarif report written to the path I've provided.

**Comment by baileympearson** [comment]
Working great for BSON: https://github.com/mongodb/js-bson/pull/695 (also tested with the driver, and non-empty reports are generated for branches where we have dismissed codeql alerts)

**Comment by blink1073** [comment]
> Working great for BSON: https://github.com/mongodb/js-bson/pull/695 (also tested with the driver, and non-empty reports are generated for branches where we have dismissed codeql alerts)

Great!  I'll give it a try with Winkerberos tomorrow

**Comment by blink1073** [comment]
I'm still getting an empty sarif file on publish, will have to do some more debugging on a fork of this branch.

**Comment by blink1073** [comment]
It works with pymongo!  I just pushed a commit that should fix CI, and made the logs print at info level.

**Review by blink1073 — APPROVED** [review]
Thank you!

**Comment by alcaeus** [comment]
Thanks for testing and fixing issues, @blink1073! Did you get it to work for a tag ref as well or only for a branch ref?

**Comment by alcaeus** [comment]
@baileympearson I've left the branch in my fork in place since you seem to be using it in js-bson. Let me know when you've changed that to `main` so I can remove the obsolete branch.
